### PR TITLE
Feat add debug draw and try to fix strange behaviors

### PR DESCRIPTION
**Pull request by moe-moe-pupil**

### CODE REVIEW

File: src/index.ejs [hunk 1]
```diff
@@ -134,6 +134,7 @@
 </head>
 
 <body>
+<canvas id="debug-canvas" width="200" height="200" style="position: absolute;top:0;right:0;z-index:100;display: none;"></canvas>
```
Comment (sloont) [inline src/index.ejs:137]: `index.ejs` only gets loaded when running taro standalone. we would have to add this html to the document that gets served from the backend

**Review by sloont — CHANGES_REQUESTED** [review]

File: ts/engine/components/physics/box2d/distsWrapper/box2dwasmWrapper.ts [hunk 2]
```diff
@@ -88,6 +87,29 @@ const box2dwasmWrapper: PhysicsDistProps = { // added by Moe'Thun for fixing mem
...
 
+		let ctx: Phaser.GameObjects.Graphics;
+		if (taro.isClient) {
+			window.taro.enableDebug = (flags = 1) => {
+				console.log('please make sure clientPhyscisEngine is not "", and enabled csp');
+				if (!component.renderer) {
+					const canvas = taro.renderer.scene.getScene('Game');
+					ctx = canvas.add.graphics().setDepth(9999);
+					ctx.setScale(30);
+					const newRenderer = new Box2dDebugDraw(box2D, new Box2dHelpers(box2D), ctx, 32).constructJSDraw();
```
Comment (sloont) [inline ts/engine/components/physics/box2d/distsWrapper/box2dwasmWrapper.ts:98]: should this line have scale of `30` instead of `32`?

**Review by sloont — CHANGES_REQUESTED** [review]

File: ts/engine/components/physics/box2d/debugDrawWrapper/box2dwasmDebugDraw.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+		this.setColorFromDebugDrawCallback(color_p);
+		const sizePixels = sizeMetres / this.canvasScaleFactor;
+		this.context.fillRect(vert.get_x() - sizePixels / 2, vert.get_y() - sizePixels / 2, sizePixels, sizePixels);
+	};
+
+	drawSegment = (vert1_p: number, vert2_p: number): void => {
+		const { wrapPointer, b2Vec2 } = this.box2D;
+		const vert1V = wrapPointer(vert1_p, b2Vec2);
+		const vert2V = wrapPointer(vert2_p, b2Vec2);
+		this.context.beginPath();
```
[the remark this answers is not in the document]
Comment (moe-moe-pupil) [reply ts/engine/components/physics/box2d/debugDrawWrapper/box2dwasmDebugDraw.ts:53]: closePath() connects the first point with the last point, it seems not necessary

File: ts/engine/components/physics/box2d/debugDrawWrapper/box2dwasmDebugDraw.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+		if (fill) {
+			this.context.fill();
+		}
+		this.context.stroke();
+
+		if (fill) {
+			//render axis marker
+			const vert2V = copyVec2(centerV);
+			vert2V.op_add(scaledVec2(axisV, radius));
+			this.context.beginPath();
```
[the remark this answers is not in the document]
Comment (moe-moe-pupil) [reply ts/engine/components/physics/box2d/debugDrawWrapper/box2dwasmDebugDraw.ts:94]: closePath() connects the first point with the last point, it seems not necessary

File: ts/engine/components/physics/box2d/debugDrawWrapper/box2dwasmDebugDraw.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	setColorFromDebugDrawCallback = (color_p: number): void => {
+		const { wrapPointer, b2Color } = this.box2D;
+		const col = wrapPointer(color_p, b2Color);
+		const red = (col.get_r() * 255) | 0;
+		const green = (col.get_g() * 255) | 0;
+		const blue = (col.get_b() * 255) | 0;
+		const colNum = red << 16 | green << 8 | blue;
+		this.context.fillStyle(colNum, 0.5);
```
[the remark this answers is not in the document]
Comment (moe-moe-pupil) [reply ts/engine/components/physics/box2d/debugDrawWrapper/box2dwasmDebugDraw.ts:38]: yep, ur right, it seems much better now,
image

File: ts/engine/components/physics/box2d/distsWrapper/box2dwasmWrapper.ts (hunk 2 above)
Comment (moe-moe-pupil) [reply ts/engine/components/physics/box2d/distsWrapper/box2dwasmWrapper.ts:98]: thanks, I changed it

**Review by sloont — APPROVED** [review]

**Review by sloont — APPROVED** [review]

**Pull request merged**